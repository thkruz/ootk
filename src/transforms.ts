/* eslint-disable init-declarations */
/**
 * @author Theodore Kruczek.
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects.
 *
 * @file The Transforms module contains a collection of conversions not contained
 * in the original SGP4 library such as ECI to ECF and ECF to RAE. This was based
 * on some of the functions in satellite.js.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Degrees, EcfVec3, EciVec3, Kilometer, LlaVec3, Radians, RaeVec3, SezVec3 } from './types';
import { PI, TAU } from './utils/constants';

class Transforms {
  public static getDegLat(radians: Radians): Degrees {
    if (radians < -PI / 2 || radians > PI / 2) {
      throw new RangeError('Latitude radians must be in range [-PI/2; PI/2].');
    }

    return Transforms.rad2deg(radians);
  }

  public static getDegLon(radians: Radians): Degrees {
    if (radians < -PI || radians > PI) {
      throw new RangeError('Longitude radians must be in range [-PI; PI].');
    }

    return Transforms.rad2deg(radians);
  }

  public static getRadLat(degrees: Degrees): Radians {
    if (degrees < -90 || degrees > 90) {
      throw new RangeError('Latitude degrees must be in range [-90; 90].');
    }

    return Transforms.deg2rad(degrees);
  }

  public static getRadLon(degrees: Degrees): Radians {
    if (degrees < -180 || degrees > 180) {
      throw new RangeError('Longitude degrees must be in range [-180; 180].');
    }

    return Transforms.deg2rad(degrees);
  }

  public static rad2deg(radians: Radians): Degrees {
    return (radians * 180) / PI;
  }

  public static deg2rad(degrees: Degrees): Radians {
    return (degrees * PI) / 180.0;
  }

  public static ecf2eci(ecf: EcfVec3, gmst: number): EciVec3 {
    /*
     * Ccar.colorado.edu/ASEN5070/handouts/coordsys.doc
     *
     * [X]     [C -S  0][X]
     * [Y]  =  [S  C  0][Y]
     * [Z]eci  [0  0  1][Z]ecf
     *
     */
    const X = ecf.x * Math.cos(gmst) - ecf.y * Math.sin(gmst);
    const Y = ecf.x * Math.sin(gmst) + ecf.y * Math.cos(gmst);
    const Z = ecf.z;

    return { x: X, y: Y, z: Z };
  }

  public static ecf2rae(lla: LlaVec3, ecf: EcfVec3): RaeVec3 {
    const sezCoords = Transforms.lla2sez(lla, ecf);

    return Transforms.sez2rae(sezCoords);
  }

  /**
   * EciToGeodetic converts eci coordinates to lla coordinates
   * @param {vec3} eci takes xyz coordinates
   * @param {number} gmst takes a number in gmst time
   * @returns {array} array containing lla coordinates
   */
  public static eci2lla(eci: EciVec3, gmst: number): LlaVec3 {
    // http://www.celestrak.com/columns/v02n03/
    const a = 6378.137;
    const b = 6356.7523142;
    const R = Math.sqrt(eci.x * eci.x + eci.y * eci.y);
    const f = (a - b) / a;
    const e2 = 2 * f - f * f;

    let lon = Math.atan2(eci.y, eci.x) - gmst;

    while (lon < -PI) {
      lon += TAU;
    }
    while (lon > PI) {
      lon -= TAU;
    }

    const kmax = 20;
    let k = 0;
    let lat = Math.atan2(eci.z, Math.sqrt(eci.x * eci.x + eci.y * eci.y));
    let C: number;

    while (k < kmax) {
      C = 1 / Math.sqrt(1 - e2 * (Math.sin(lat) * Math.sin(lat)));
      lat = Math.atan2(eci.z + a * C * e2 * Math.sin(lat), R);
      k += 1;
    }
    const alt = R / Math.cos(lat) - a * C;

    return { lon: <Radians>lon, lat: <Radians>lat, alt: <Kilometer>alt };
  }

  public static eci2ecf(eci: EciVec3, gmst: number): EcfVec3 {
    /*
     * Ccar.colorado.edu/ASEN5070/handouts/coordsys.doc
     *
     * [X]     [C -S  0][X]
     * [Y]  =  [S  C  0][Y]
     * [Z]eci  [0  0  1][Z]ecf
     *
     *
     * Inverse:
     * [X]     [C  S  0][X]
     * [Y]  =  [-S C  0][Y]
     * [Z]ecf  [0  0  1][Z]eci
     */

    const x = eci.x * Math.cos(gmst) + eci.y * Math.sin(gmst);
    const y = eci.x * -Math.sin(gmst) + eci.y * Math.cos(gmst);
    const { z } = eci;

    return {
      x,
      y,
      z,
    };
  }

  public static lla2ecf(lla: LlaVec3): EcfVec3 {
    const { lon, lat, alt } = lla;

    const a = 6378.137;
    const b = 6356.7523142;
    const f = (a - b) / a;
    const e2 = 2 * f - f * f;
    const normal = a / Math.sqrt(1 - e2 * (Math.sin(lat) * Math.sin(lat)));

    const x = (normal + alt) * Math.cos(lat) * Math.cos(lon);
    const y = (normal + alt) * Math.cos(lat) * Math.sin(lon);
    const z = (normal * (1 - e2) + alt) * Math.sin(lat);

    return {
      x: <Kilometer>x,
      y: <Kilometer>y,
      z: <Kilometer>z,
    };
  }

  public static lla2sez(lla: LlaVec3, ecf: EcfVec3): SezVec3 {
    /*
     * http://www.celestrak.com/columns/v02n02/
     * TS Kelso's method, except I'm using ECF frame
     * and he uses ECI.
     */

    const { lon, lat } = lla;

    const observerEcf = Transforms.lla2ecf(lla);

    const rx = ecf.x - observerEcf.x;
    const ry = ecf.y - observerEcf.y;
    const rz = ecf.z - observerEcf.z;

    // Top is short for topocentric
    const south = Math.sin(lat) * Math.cos(lon) * rx + Math.sin(lat) * Math.sin(lon) * ry - Math.cos(lat) * rz;

    const east = -Math.sin(lon) * rx + Math.cos(lon) * ry;

    const zenith = Math.cos(lat) * Math.cos(lon) * rx + Math.cos(lat) * Math.sin(lon) * ry + Math.sin(lat) * rz;

    return { s: <Kilometer>south, e: <Kilometer>east, z: <Kilometer>zenith };
  }

  public static rae2sez(rae: RaeVec3): SezVec3 {
    // Az,el,range to sez convertion
    const south = -rae.rng * Math.cos(rae.el) * Math.cos(rae.az);
    const east = rae.rng * Math.cos(rae.el) * Math.sin(rae.az);
    const zenith = rae.rng * Math.sin(rae.el);

    return {
      s: <Kilometer>south,
      e: <Kilometer>east,
      z: <Kilometer>zenith,
    };
  }

  public static rae2ecf(rae: RaeVec3, lla: LlaVec3): EcfVec3 {
    const obsEcf = Transforms.lla2ecf(lla);
    const sez = Transforms.rae2sez(rae);

    // Some needed calculations
    const slat = Math.sin(lla.lat);
    const slon = Math.sin(lla.lon);
    const clat = Math.cos(lla.lat);
    const clon = Math.cos(lla.lon);

    const x = slat * clon * sez.s + -slon * sez.e + clat * clon * sez.z + obsEcf.x;
    const y = slat * slon * sez.s + clon * sez.e + clat * slon * sez.z + obsEcf.y;
    const z = -clat * sez.s + slat * sez.z + obsEcf.z;

    return { x: <Kilometer>x, y: <Kilometer>y, z: <Kilometer>z };
  }

  /**
   * @param {Object} sez Containing SEZ coordinates
   * @param {Number} sez.s Positive horizontal vector S due south.
   * @param {Number} sez.e Positive horizontal vector E due east.
   * @param {Number} sez.z Vector Z normal to the surface of the earth (up).
   * @returns {Object} Rng, Az, El array
   */
  public static sez2rae(sez: SezVec3): RaeVec3 {
    const rng = Math.sqrt(sez.s * sez.s + sez.e * sez.e + sez.z * sez.z);
    const el = Math.asin(sez.z / rng);
    const az = Math.atan2(-sez.e, sez.s) + PI;

    return {
      rng: <Kilometer>rng,
      az: <Radians>az,
      el: <Radians>el,
    };
  }
}

export { Transforms };
