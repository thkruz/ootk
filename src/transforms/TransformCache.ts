/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export class TransformCache {
  private static cache_: Map<string, unknown> = new Map();

  static get(key: string): unknown {
    const value = this.cache_.get(key);

    return value;
  }
  static add(key: string, value: unknown) {
    this.cache_.set(key, value);

    // Max of 1000 items in the cache
    if (this.cache_.size > 1000) {
      const firstKey = this.cache_.keys().next().value;

      this.cache_.delete(firstKey);
    }
  }
}
