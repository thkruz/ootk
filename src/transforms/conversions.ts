/**
 * @author Theodore Kruczek.
 * @license MIT
 * @copyright (c) 2022-2025 Theodore Kruczek Permission is
 * hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { DEG2RAD, Degrees, PI, RAD2DEG, Radians } from '../main.js';

/**
 * Converts radians to degrees.
 * @param radians The value in radians to be converted.
 * @returns The value in degrees.
 */
export function rad2deg(radians: Radians): Degrees {
  return <Degrees>((radians * 180) / PI);
}

/**
 * Converts degrees to radians.
 * @param degrees The value in degrees to be converted.
 * @returns The value in radians.
 */
export function deg2rad(degrees: Degrees): Radians {
  return <Radians>((degrees * PI) / 180.0);
}

/**
 * Converts radians to degrees latitude.
 * @param radians The radians value to convert.
 * @returns The corresponding degrees latitude.
 * @throws RangeError if the radians value is outside the range [-PI/2; PI/2].
 */
export function getDegLat(radians: Radians): Degrees {
  if (radians < -PI / 2 || radians > PI / 2) {
    throw new RangeError('Latitude radians must be in range [-PI/2; PI/2].');
  }

  return (radians * RAD2DEG) as Degrees;
}

/**
 * Converts radians to degrees for longitude.
 * @param radians The value in radians to be converted.
 * @returns The converted value in degrees.
 * @throws {RangeError} If the input radians is not within the range [-PI; PI].
 */
export function getDegLon(radians: Radians): Degrees {
  if (radians < -PI || radians > PI) {
    throw new RangeError('Longitude radians must be in range [-PI; PI].');
  }

  return (radians * RAD2DEG) as Degrees;
}

/**
 * Converts degrees to radians for latitude.
 * @param degrees The degrees value to convert.
 * @returns The equivalent radians value.
 * @throws {RangeError} If the degrees value is not within the range [-90, 90].
 */
export function getRadLat(degrees: Degrees): Radians {
  if (degrees < -90 || degrees > 90) {
    throw new RangeError('Latitude degrees must be in range [-90; 90].');
  }

  return (degrees * DEG2RAD) as Radians;
}

/**
 * Converts degrees to radians.
 * @param degrees The value in degrees to be converted.
 * @returns The value in radians.
 * @throws {RangeError} If the input degrees are not within the range [-180; 180].
 */
export function getRadLon(degrees: Degrees): Radians {
  if (degrees < -180 || degrees > 180) {
    throw new RangeError('Longitude degrees must be in range [-180; 180].');
  }

  return (degrees * DEG2RAD) as Radians;
}
