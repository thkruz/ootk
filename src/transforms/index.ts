export * from './conversions.js';
export * from './transforms.js';
