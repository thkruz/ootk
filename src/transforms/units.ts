import { Degrees, Radians } from '@src/ootk';
import { DEG2RAD, PI, RAD2DEG } from '@src/utils/constants';

/**
 * Converts radians to degrees.
 * @param radians The value in radians to be converted.
 * @returns The value in degrees.
 */
export function rad2deg(radians: Radians): Degrees {
  return <Degrees>((radians * 180) / PI);
}

/**
 * Converts degrees to radians.
 * @param degrees The value in degrees to be converted.
 * @returns The value in radians.
 */
export function deg2rad(degrees: Degrees): Radians {
  return <Radians>((degrees * PI) / 180.0);
}

/**
 * Converts radians to degrees latitude.
 * @param radians The radians value to convert.
 * @returns The corresponding degrees latitude.
 * @throws RangeError if the radians value is outside the range [-PI/2; PI/2].
 */
export function getDegLat(radians: Radians): Degrees {
  if (radians < -PI / 2 || radians > PI / 2) {
    throw new RangeError('Latitude radians must be in range [-PI/2; PI/2].');
  }

  return (radians * RAD2DEG) as Degrees;
}

/**
 * Converts radians to degrees for longitude.
 * @param radians The value in radians to be converted.
 * @returns The converted value in degrees.
 * @throws {RangeError} If the input radians is not within the range [-PI; PI].
 */
export function getDegLon(radians: Radians): Degrees {
  if (radians < -PI || radians > PI) {
    throw new RangeError('Longitude radians must be in range [-PI; PI].');
  }

  return (radians * RAD2DEG) as Degrees;
}

/**
 * Converts degrees to radians for latitude.
 * @param degrees The degrees value to convert.
 * @returns The equivalent radians value.
 * @throws {RangeError} If the degrees value is not within the range [-90, 90].
 */
export function getRadLat(degrees: Degrees): Radians {
  if (degrees < -90 || degrees > 90) {
    throw new RangeError('Latitude degrees must be in range [-90; 90].');
  }

  return (degrees * DEG2RAD) as Radians;
}

/**
 * Converts degrees to radians.
 * @param degrees The value in degrees to be converted.
 * @returns The value in radians.
 * @throws {RangeError} If the input degrees are not within the range [-180; 180].
 */
export function getRadLon(degrees: Degrees): Radians {
  if (degrees < -180 || degrees > 180) {
    throw new RangeError('Longitude degrees must be in range [-180; 180].');
  }

  return (degrees * DEG2RAD) as Radians;
}
