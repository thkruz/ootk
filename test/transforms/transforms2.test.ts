import {
  Degrees,
  ecf2eci,
  ecf2enu,
  ecf2rae,
  eci2ecf,
  eci2lla,
  getDegLat,
  getDegLon,
  getRadLat,
  getRadLon,
  Kilometers,
  lla2ecf,
  Radians,
  rae2ecf,
  rae2enu,
  rae2sez,
  Sensor,
  eci2rae,
  Vec3,
  lla2eci,
  calcGmst,
  lla2ecef,
  rae2eci,
} from '../../src/main';
import { transformsData } from '../transforms/transformsData';

const numDigits = 6;

describe('Latitude & longitude conversions', () => {
  const {
    validLatitudes,
    validLongitudes,
    validGeodeticToEcf,
    validEciToGeodetic,
    validEciToEcf,
    validEcfToEci,
    validEcfToLookangles,
    invalidLatitudes,
    invalidLongitudes,
  } = transformsData;

  validLatitudes.forEach((item) => {
    it(`convert valid latitude value (${item.radians} radians) to degrees`, () => {
      expect(getDegLat(item.radians)).toBeCloseTo(item.degrees, numDigits);
    });
    it(`convert valid latitude value (${item.degrees} degrees) to radians`, () => {
      expect(getRadLat(item.degrees)).toBeCloseTo(item.radians, numDigits);
    });
  });

  validLongitudes.forEach((item) => {
    it(`convert valid longitude value (${item.radians} radians) to degrees`, () => {
      expect(getDegLon(item.radians)).toBeCloseTo(item.degrees, numDigits);
    });
    it(`convert valid longitude value (${item.degrees} degrees) to radians`, () => {
      expect(getRadLon(item.degrees)).toBeCloseTo(item.radians, numDigits);
    });
  });

  validGeodeticToEcf.forEach((item) => {
    it('convert valid LLA coordinates to ECF', () => {
      const ecfCoordinates = lla2ecf(item.lla);

      expect(ecfCoordinates.x).toBeCloseTo(item.ecf.x);
      expect(ecfCoordinates.y).toBeCloseTo(item.ecf.y);
      expect(ecfCoordinates.z).toBeCloseTo(item.ecf.z);
    });
  });

  validEciToGeodetic.forEach((item) => {
    it('convert valid ECI coordinates to LLA', () => {
      const llaCoordinates = eci2lla(item.eci, item.gmst);

      expect(llaCoordinates.lon).toBeCloseTo(item.lla.lon);
      expect(llaCoordinates.lat).toBeCloseTo(item.lla.lat);
      expect(llaCoordinates.alt).toBeCloseTo(item.lla.alt);
    });
  });

  validEciToEcf.forEach((item) => {
    it('convert valid ECI coordinates to ECF', () => {
      const ecfCoordinates = eci2ecf(item.eci, item.gmst);

      expect(ecfCoordinates.x).toBeCloseTo(item.ecf.x);
      expect(ecfCoordinates.y).toBeCloseTo(item.ecf.y);
      expect(ecfCoordinates.z).toBeCloseTo(item.ecf.z);
    });
  });

  validEcfToEci.forEach((item) => {
    it('convert valid ECF coordinates to ECI', () => {
      const eciCoordinates = ecf2eci(item.ecf, item.gmst);

      expect(eciCoordinates.x).toBeCloseTo(item.eci.x);
      expect(eciCoordinates.y).toBeCloseTo(item.eci.y);
      expect(eciCoordinates.z).toBeCloseTo(item.eci.z);
    });
  });

  validEcfToLookangles.forEach((item) => {
    it('convert valid ECF coordinates to RAE', () => {
      const raeCoordinates = ecf2rae(item.lla, item.satelliteEcf);

      expect(raeCoordinates.rng).toBeCloseTo(item.rae.rng, 0);
      expect(raeCoordinates.az).toBeCloseTo(item.rae.az, 1);
      expect(raeCoordinates.el).toBeCloseTo(item.rae.el, 1);
    });
  });

  invalidLatitudes.forEach((item) => {
    it(`convert invalid latitude value (${item.radians} radians) to degrees`, () => {
      expect(() => getDegLat(item.radians)).toThrowError(RangeError);
    });
    it(`convert invalid latitude value (${item.degrees} degrees) to radians`, () => {
      expect(() => getRadLat(item.degrees)).toThrowError(RangeError);
    });
  });

  invalidLongitudes.forEach((item) => {
    it(`convert invalid longitude value (${item.radians} radians) to degrees`, () => {
      expect(() => getDegLon(item.radians)).toThrowError(RangeError);
    });
    it(`convert invalid longitude value (${item.degrees} degrees) to radians`, () => {
      expect(() => getRadLon(item.degrees)).toThrowError(RangeError);
    });
  });
});

describe('Rae2Sez', () => {
  it('should convert valid RAE coordinates to SEZ', () => {
    const { rae, sez } = transformsData.validRae2Sez[0] as {
      rae: {
        rng: Kilometers;
        az: Radians;
        el: Radians;
      }
      sez: {
        s: number;
        e: number;
        z: number;
      };
    };

    const sezCoordinates = rae2sez(rae);

    expect(sezCoordinates.s).toBeCloseTo(sez.s);
    expect(sezCoordinates.e).toBeCloseTo(sez.e);
    expect(sezCoordinates.z).toBeCloseTo(sez.z);
  });
});

describe('Rae2Ecf', () => {
  it('should convert valid RAE coordinates to ECF', () => {
    // const { rae, ecf, lla } = transformData.validRae2Ecf[0];
    const ecf = {
      x: 4000,
      y: 4000,
      z: 4000,
    };
    const lla = {
      lon: 0 as Degrees,
      lat: 0 as Degrees,
      alt: 0 as Kilometers,
    };
    const rae = ecf2rae(lla, ecf);

    const ecfCoordinates = rae2ecf(rae, lla);

    expect(ecfCoordinates.x).toBeCloseTo(ecf.x);
    expect(ecfCoordinates.y).toBeCloseTo(ecf.y);
    expect(ecfCoordinates.z).toBeCloseTo(ecf.z);
  });

  // ecf2enu
  it('should convert valid ECF coordinates to ENU', () => {
    const ecf = {
      x: 4000,
      y: 4000,
      z: 4000,
    };
    const lla = {
      lon: 0 as Degrees,
      lat: 0 as Degrees,
      alt: 0 as Kilometers,
    };

    const enuCoordinates = ecf2enu(ecf, lla);

    expect(enuCoordinates).toMatchSnapshot();
  });

  // enu2rf
  it('should convert valid ENU coordinates to RF', () => {
    const ecf = {
      x: 4000,
      y: 4000,
      z: 4000,
    };
    const lla = {
      lon: 0 as Degrees,
      lat: 0 as Degrees,
      alt: 0 as Kilometers,
    };

    const enuCoordinates = ecf2enu(ecf, lla);

    expect(enuCoordinates).toMatchSnapshot();
  });

  // lla2eci
  it('should convert valid LLA coordinates to ECI', () => {
    const lla = {
      lon: 0 as Radians,
      lat: 0 as Radians,
      alt: 0 as Kilometers,
    };
    const exampleDate = new Date(1705109326817);
    const { gmst } = calcGmst(exampleDate);
    const eciCoordinates = lla2eci(lla, gmst);

    expect(eciCoordinates).toMatchSnapshot();
  });

  // lla2ecef
  it('should convert valid LLA coordinates to ECF', () => {
    const lla = {
      lon: 0 as Degrees,
      lat: 0 as Degrees,
      alt: 0 as Kilometers,
    };
    const ecfCoordinates = lla2ecef(lla);

    expect(ecfCoordinates).toMatchSnapshot();
  });

  // rae2eci
  it('should convert valid RAE coordinates to ECI', () => {
    const rae = {
      rng: 0 as Kilometers,
      az: 0 as Degrees,
      el: 0 as Degrees,
    };
    const sensor = new Sensor({
      lat: 0 as Degrees,
      lon: 0 as Degrees,
      alt: 0 as Kilometers,
      minAz: 0 as Degrees,
      maxAz: 0 as Degrees,
      minEl: 0 as Degrees,
      maxEl: 0 as Degrees,
      minRng: 0 as Kilometers,
      maxRng: 0 as Kilometers,
    }) as Sensor;
    const exampleDate = new Date(1705109326817);
    const { gmst } = calcGmst(exampleDate);

    const eciCoordinates = rae2eci(rae, sensor, gmst);

    expect(eciCoordinates).toMatchSnapshot();
  });

  // rae2enu
  it('should convert valid RAE coordinates to ENU', () => {
    const rae = {
      rng: 0 as Kilometers,
      az: 0 as Degrees,
      el: 0 as Degrees,
    };
    const enuCoordinates = rae2enu(rae);

    expect(enuCoordinates).toMatchSnapshot();
  });

  // eci2rae
  it('should convert valid ECI coordinates to RAE', () => {
    const eci = {
      x: 4000,
      y: 4000,
      z: 4000,
    } as Vec3<Kilometers>;
    const sensor = new Sensor({
      lat: 0 as Degrees,
      lon: 0 as Degrees,
      alt: 0 as Kilometers,
      minAz: 0 as Degrees,
      maxAz: 0 as Degrees,
      minEl: 0 as Degrees,
      maxEl: 0 as Degrees,
      minRng: 0 as Kilometers,
      maxRng: 0 as Kilometers,
    }) as Sensor;

    const exampleDate = new Date(1705109326817);
    const raeCoordinates = eci2rae(exampleDate, eci, sensor);

    expect(raeCoordinates).toMatchSnapshot();
  });
});
