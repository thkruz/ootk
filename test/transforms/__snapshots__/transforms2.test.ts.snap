// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rae2Ecf should convert valid ECF coordinates to ENU 1`] = `
Object {
  "x": 4000,
  "y": 4000,
  "z": 4000,
}
`;

exports[`Rae2Ecf should convert valid ECI coordinates to RAE 1`] = `
Object {
  "az": 305.2667894949389,
  "el": -42.29513998190954,
  "rng": 9365.828830438693,
}
`;

exports[`Rae2Ecf should convert valid ENU coordinates to RF 1`] = `
Object {
  "x": 4000,
  "y": 4000,
  "z": 4000,
}
`;

exports[`Rae2Ecf should convert valid LLA coordinates to ECF 1`] = `
Object {
  "x": 6378.137,
  "y": 0,
  "z": 0,
}
`;

exports[`Rae2Ecf should convert valid LLA coordinates to ECI 1`] = `
Object {
  "x": -4444.532338535956,
  "y": 4564.633188736234,
  "z": 0,
}
`;

exports[`Rae2Ecf should convert valid RAE coordinates to ECI 1`] = `
Object {
  "x": -4449.505609609978,
  "y": 4569.740848321526,
  "z": 0,
}
`;

exports[`Rae2Ecf should convert valid RAE coordinates to ENU 1`] = `
Object {
  "x": 0,
  "y": 0,
  "z": 0,
}
`;
